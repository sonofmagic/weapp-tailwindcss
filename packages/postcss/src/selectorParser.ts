export { isOnlyBeforeAndAfterPseudoElement } from './selectorParser/before-after'

export { getFallbackRemove } from './selectorParser/fallback'

export { ruleTransformSync } from './selectorParser/rule-transformer'

export {
  composeIsPseudoAst,
  getCombinatorSelectorAst,
  mklist,
} from './selectorParser/utils'
