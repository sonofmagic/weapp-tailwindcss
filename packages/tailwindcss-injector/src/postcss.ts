import type { Input, PluginCreator } from 'postcss'
import type { Config } from 'tailwindcss'
import type { Options } from './types'
import postcss from 'postcss'
import tailwindcss from 'tailwindcss'
import { loadConfig } from 'tailwindcss-config'
import { getConfig } from './config'
import { postcssPlugin } from './constants'
import { regExpTest, removeFileExtension } from './utils'
import { getDepFiles } from './wxml'

const configCache = new Map<string, Promise<Config | undefined>>()

async function loadTailwindConfigOnce(cwd: string, configPath: string) {
  const cacheKey = `${cwd}::${configPath}`
  let cached = configCache.get(cacheKey)
  if (!cached) {
    cached = loadConfig({
      cwd,
      config: configPath,
    }).then(result => result?.config as Config | undefined)
    configCache.set(cacheKey, cached)
  }
  return cached
}

function cloneConfig(config: Config | undefined): Config | undefined {
  if (!config) {
    return config
  }

  return {
    ...config,
  }
}

async function resolveTailwindConfig(config: Options['config'], cwd: string, sourceInput: Input | undefined) {
  if (!config) {
    return undefined
  }

  const resolved = typeof config === 'function' ? config(sourceInput) : config

  if (typeof resolved === 'string') {
    const loaded = await loadTailwindConfigOnce(cwd, resolved)
    return cloneConfig(loaded)
  }

  return cloneConfig(resolved as Config)
}

// function isObject(obj: any): obj is object {
//   return typeof obj === 'object' && obj !== null
// }
export type {
  Options,
}

const creator: PluginCreator<Partial<Options>> = (options) => {
  const { config, filter, directiveParams, cwd, extensions, insertAfterAtRulesNames, insertAfterComments } = getConfig(options)

  const extensionsGlob = `{${extensions.join(',')}}`

  return {
    postcssPlugin,
    plugins: [
      {
        postcssPlugin: `${postcssPlugin}:post`,
        async Once(root, helpers) {
          const sourceInput = root.source?.input
          if (filter(sourceInput)) {
            const nodes = root.nodes ?? []
            let atruleAnchorIndex = -1
            let commentAnchorIndex = -1
            const directivePresence = new Set<string>()

            nodes.forEach((node, index) => {
              if (node.type === 'atrule') {
                const nameMatches = insertAfterAtRulesNames.includes(node.name)
                if (nameMatches) {
                  atruleAnchorIndex = Math.max(atruleAnchorIndex, index)
                }
                if (typeof node.params === 'string') {
                  directivePresence.add(node.params)
                }
              }
              else if (node.type === 'comment' && regExpTest(insertAfterComments, node.text)) {
                commentAnchorIndex = Math.max(commentAnchorIndex, index)
              }
            })

            const anchorIndex = Math.max(atruleAnchorIndex, commentAnchorIndex)
            const anchorNode = anchorIndex >= 0 ? nodes[anchorIndex] : undefined
            let lastInserted = anchorNode

            for (const params of directiveParams) {
              if (!directivePresence.has(params)) {
                const atRule = helpers.atRule({ name: 'tailwind', params })
                if (lastInserted) {
                  root.insertAfter(lastInserted, atRule)
                }
                else {
                  root.prepend(atRule)
                }
                lastInserted = atRule
                directivePresence.add(params)
              }
            }

            if (sourceInput?.file) {
              const tailwindcssConfig = (await resolveTailwindConfig(config, cwd, sourceInput)) ?? { content: [] }

              const basename = removeFileExtension(sourceInput.file)
              const contentEntries = [`${basename}.${extensionsGlob}`]

              const deps = await getDepFiles(`${basename}.wxml`)
              for (const dep of deps) {
                contentEntries.push(`${removeFileExtension(dep)}.${extensionsGlob}`)
              }

              const uniqueEntries = Array.from(new Set(contentEntries))
              const { content: existingContent } = tailwindcssConfig as { content?: Config['content'] }

              if (!existingContent || Array.isArray(existingContent)) {
                tailwindcssConfig.content = uniqueEntries
              }
              else if (typeof existingContent === 'object' && existingContent !== null) {
                const normalizedContent = { ...existingContent }
                const currentFiles = normalizedContent.files
                const normalizedFiles = Array.isArray(currentFiles)
                  ? currentFiles
                  : typeof currentFiles === 'string'
                    ? [currentFiles]
                    : []
                normalizedContent.files = Array.from(new Set([...normalizedFiles, ...uniqueEntries]))
                tailwindcssConfig.content = normalizedContent
              }
              else {
                tailwindcssConfig.content = uniqueEntries
              }

              await postcss([
                tailwindcss(tailwindcssConfig),
              ]).process(root, {
                from: sourceInput.file,
              }).async()
            }
          }
        },
      },
    ],
  }
}
creator.postcss = true

export default creator
