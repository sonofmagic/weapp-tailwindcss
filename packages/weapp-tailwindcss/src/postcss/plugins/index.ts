import type { AcceptedPlugin } from 'postcss'
// @ts-expect-error
import postcssIsPseudoClass from '@csstools/postcss-is-pseudo-class'
import postcssRem2rpx from 'postcss-rem-to-responsive-pixel'
import type { IStyleHandlerOptions } from '../../types'
import { postcssWeappTailwindcssPrePlugin } from './pre'
import { postcssWeappTailwindcssPostPlugin } from './post'
import { createContext } from './ctx'

export function getPlugins(options: IStyleHandlerOptions) {
  const ctx = createContext()
  options.ctx = ctx
  const plugins: AcceptedPlugin[] = [
    ...(options.postcssOptions?.plugins ?? []),
    postcssWeappTailwindcssPrePlugin(options),
    postcssIsPseudoClass({
      specificityMatchingName: 'weapp-tw-ig',
    }),
  ]
  if (options.rem2rpx) {
    plugins.push(
      postcssRem2rpx(
        typeof options.rem2rpx === 'object'
          ? options.rem2rpx
          : {
              rootValue: 32,
              propList: ['*'],
              transformUnit: 'rpx',
            },
      ),
    )
  }
  plugins.push(postcssWeappTailwindcssPostPlugin(options))
  return plugins
}

export { postcssWeappTailwindcssPostPlugin } from './post'
export { postcssWeappTailwindcssPrePlugin } from './pre'
// @ts-expect-error
export { default as postcssIsPseudoClass } from '@csstools/postcss-is-pseudo-class'
export { default as postcssRem2rpx } from 'postcss-rem-to-responsive-pixel'
