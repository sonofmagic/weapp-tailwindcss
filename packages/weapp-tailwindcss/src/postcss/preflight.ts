import type { CssPreflightOptions, IPropValue } from '../types'

export type InjectPreflight = () => IPropValue[]

export function createInjectPreflight(options?: CssPreflightOptions): InjectPreflight {
  const result: IPropValue[] = []
  // if options false ,do no thing
  if (options && typeof options === 'object') {
    const entries = Object.entries(options)
    for (const [prop, value] of entries) {
      if (value !== false) {
        result.push({
          prop,
          value: value.toString(),
        })
      }
    }
  }

  return () => {
    return result
  }
}
