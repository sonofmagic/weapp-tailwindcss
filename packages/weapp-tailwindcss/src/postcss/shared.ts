import { SimpleMappingChars2String, escape } from '../escape'
import type { InternalCssSelectorReplacerOptions } from '../types'
// css 中，要多加一个 '\' 来转义
// for raw css selector
// export function cssSelectorReplacer(selector: string, escapeEntries = MappingChars2StringEntries) {
//   return escape(selector, true, escapeEntries).replace(/\\2c /g, dic[','])
// }

export function internalCssSelectorReplacer(
  selectors: string,
  options: InternalCssSelectorReplacerOptions = {
    escapeMap: SimpleMappingChars2String,
  },
) {
  const { mangleContext, escapeMap } = options
  if (mangleContext) {
    selectors = mangleContext.cssHandler(selectors)
  }
  return escape(selectors, {
    map: escapeMap,
  })
}

export function composeIsPseudo(strs: string | string[]) {
  if (typeof strs === 'string') {
    return strs
  }
  if (strs.length > 1) {
    return `:is(${strs.join(',')})`
  }
  return strs.join('')
}
