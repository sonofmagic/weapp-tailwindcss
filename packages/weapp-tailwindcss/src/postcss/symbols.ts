export const VariablesScopeSymbol = Symbol('VariablesScope')
