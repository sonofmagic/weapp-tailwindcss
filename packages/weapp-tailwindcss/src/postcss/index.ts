import postcss from 'postcss'
import type { IStyleHandlerOptions } from '../types'
import { defuOverrideArray } from '../utils'
import { getPlugins } from './plugins'

export function styleHandler(rawSource: string, options: IStyleHandlerOptions) {
  return postcss(getPlugins(options))
    .process(
      rawSource,
      options.postcssOptions?.options ?? {
        from: undefined,
      },
    )
    .async()
}

export function createStyleHandler(options: Partial<IStyleHandlerOptions>) {
  return (rawSource: string, opt?: Partial<IStyleHandlerOptions>) => {
    return styleHandler(
      rawSource,
      defuOverrideArray<IStyleHandlerOptions, Partial<IStyleHandlerOptions>[]>(opt as IStyleHandlerOptions, options),
    )
  }
}
