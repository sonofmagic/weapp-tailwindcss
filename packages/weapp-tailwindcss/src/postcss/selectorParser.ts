import selectorParser from 'postcss-selector-parser'
import type { SyncProcessor } from 'postcss-selector-parser'
import type { Rule } from 'postcss'
import { composeIsPseudo, internalCssSelectorReplacer } from './shared'
import type { IStyleHandlerOptions } from '@/types'

const createRuleTransform = (rule: Rule, options: IStyleHandlerOptions) => {
  const { escapeMap, mangleContext, cssSelectorReplacement, cssRemoveHoverPseudoClass } = options

  const transform: SyncProcessor = (selectors) => {
    selectors.walk((selector) => {
      // do something with the selector
      // node.selector.replace(/\*/g, 'view')
      if (selector.type === 'universal' && cssSelectorReplacement && cssSelectorReplacement.universal) {
        selector.value = composeIsPseudo(cssSelectorReplacement.universal)
      }

      if (cssRemoveHoverPseudoClass && selector.type === 'selector') {
        const node = selector.nodes.find((x) => x.type === 'pseudo' && x.value === ':hover')
        node && selector.remove()
      }

      if (
        selector.type === 'pseudo' &&
        selector.value === ':root' &&
        cssSelectorReplacement &&
        cssSelectorReplacement.root
      ) {
        selector.value = composeIsPseudo(cssSelectorReplacement.root)
      }

      if (selector.type === 'class') {
        selector.value = internalCssSelectorReplacer(selector.value, {
          escapeMap,
          mangleContext
        })
      }
    })
    if (selectors.length === 0) {
      rule.remove()
    }
  }
  return transform
}

const getRuleTransformer = (rule: Rule, options: IStyleHandlerOptions) => {
  return selectorParser(createRuleTransform(rule, options))
}

export const ruleTransformSync = (rule: Rule, options: IStyleHandlerOptions) => {
  const transformer = getRuleTransformer(rule, options)

  return transformer.transformSync(rule, {
    lossless: false,
    updateSelector: true
  })
}

export function isOnlyBeforeAndAfterPseudoElement(node: Rule) {
  let b = false
  let a = false

  selectorParser((selectors) => {
    selectors.walkPseudos((s) => {
      if (s.parent?.length === 1) {
        if (/^:?:before$/.test(s.value)) {
          b = true
        }
        if (/^:?:after$/.test(s.value)) {
          a = true
        }
      }
    })
  }).astSync(node)

  return b && a
}

export const fallbackRemove = selectorParser((selectors) => {
  selectors.walk((selector) => {
    if (selector.type === 'universal') {
      selector.parent?.remove()
    }
    if (selector.type === 'pseudo' && selector.value === ':is') {
      selector.parent?.remove()
    }
  })
})
