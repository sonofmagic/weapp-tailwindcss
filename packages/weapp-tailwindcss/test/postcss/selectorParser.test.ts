import { fallbackRemove } from '@/postcss/selectorParser'

describe('selectorParser', () => {
  it('fallbackRemove case 0', async () => {
    const x = await fallbackRemove.process('#app-provider :is(.space-x-4>view+view)')
    expect(x).toMatchSnapshot()
  })

  it('fallbackRemove case 1', async () => {
    const x = await fallbackRemove.process('#app :is(.space-x-0>view+view)')
    expect(x).toMatchSnapshot()
  })
})
