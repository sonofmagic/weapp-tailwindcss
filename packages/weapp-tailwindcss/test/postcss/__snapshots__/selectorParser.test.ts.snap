// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`selectorParser > fallbackRemove case 0 1`] = `"#app-provider .space-x-4>view+view"`;

exports[`selectorParser > fallbackRemove case 1 1`] = `"#app .space-x-0>view+view"`;
